### ISSUE TITLE
Failed dependency resolution on scrooge_scala_library

### ISSUE THREAD

**Issue by gkossakowski** [bazelbuild/intellij#207]
The latest Bazel plugin has troubles resolving dependencies (e.g. generated classes) coming out of `scrooge_scala_library`. I've put a small repro of the problem available at https://github.com/gkossakowski/bazel-scala-example/commit/76a2eac0277fd94b94ee0278ee14757d6a84ec61

Steps to reproduce:

1. Import that repo with the bazel intellij plugin
2. Open `Xyz.scala` file and see `example.thrift.SomeThrift` marked as red: intellij doesn't know where to find the class file for it.

Notes:
  * this issue is non-deterministic, it rarely shows up for me in this small repo, always shows up in our large internal repo at Stripe. Some people internally could reproduce it and others couldn't
  * we found out that changing the dependency from `"//example-thrift"` to `"//example-thrift:example-thrift"` makes the issue go away which is the workaround we might use. We're surprised by this behavior as the two target names are equivalent in bazel

Tested with IntelliJ 2017.3 and Bazel plugin 2017.11.20.0.4. Discovered during Stripe's migration to the plugin. The non-determinism makes us scratch our heads.

**Comment by gkossakowski** [comment]
The workaround is a bit hard to use because the latest buildifier enforces the short-label syntax.

**Comment by brendandouglas** [comment]
Thanks for the repro steps, I see the same label format bug.

The Bazel plugin gets all its dependency information from the Bazel invocations, so whatever bug is causing that workaround to have an effect is coming from Bazel.

Might be worth asking the Bazel folks directly?

**Comment by brendandouglas** [comment]
Actually, any change whatsoever to that BUILD file (followed by a resync) seems to fix this (e.g. adding a newline to the end of the file).

I'll take a closer look.

**Comment by gkossakowski** [comment]
Thanks for a quick response. I think the newline change fixing this issue is part of the non-determinism.

**Comment by brendandouglas** [comment]
It is a plugin-side issue. The situation:
- we generally rely on jdeps information to determine the minimum compile-time classpath. Only jars listed in jdeps are actually added to the project libraries
- one exception to this: for targets in the 'working set' (list of modified files), we add libraries for all dependencies, not just those required at compile time (as determined by the jdeps data)

Now, we're not getting any jdeps data from the targets in this project (perhaps this is broken for scala rules?), so the ':example-thrift' dependency is generally ignored, unless ':example-lib-bar' is in the working set (e.g. because it's been locally modified, by changing the dep target format, or adding whitespace).

Next step is to find out why jdeps information isn't available here. Perhaps the scala rule handling in our aspect has regressed, or perhaps the scala rules aren't supplying jdeps via java_common.provider.

**Comment by gkossakowski** [comment]
scala rules create a provider with `java_common.create_provider`. Will provider created that way carry jdeps attribute? 
Here's the code: https://github.com/bazelbuild/rules_scala/blob/master/scala/scala.bzl#L605

Does intellij add a jar to project libraries only for targets whose sources are not in the source tree?

**Comment by gkossakowski** [comment]
Thanks for solving the mystery why touching a file makes this issue go away! Good to hear that non-determinism was a misobservation. What's the reason for treating targets in working set in a special way?

**Comment by brendandouglas** [comment]
Restricting libraries to the minimal compile-time classpath for working set deps wasn't a great experience. Users would add a dep to their BUILD file, re-sync, and expect it to show up in auto-complete results, but of course that won't happen until the dep is added to the project libraries (requiring the user to manually type in the import string the first time they add a dependency.

On the other hand, adding unused deps to the project libraries for all targets destroys performance.

The working set logic tries to find a balance between the two.

As far as exposing jdeps information via java_common, there's no way to do that yet with the 'create_provider' API (java_common.provider is still a WIP, and doesn't provide all the data needed by an IDE). The compile action (java_common.compile) will expose jdeps information, so if possible you might want to look at using that when compiling.

Otherwise, can you file a bazel-side bug, mentioning that the java_common API work is causing issues? (it might help to bump the priority)

**Comment by gkossakowski** [comment]
Thanks for the backstory. I almost get it. 
>but of course that won't happen until the dep is added to the project libraries (requiring the user to manually type in the import string the first time they add a dependency.

Maybe I misunderstand what project libraries are. I thought we were talking about external dependencies fetched by `maven_jar` or something similar. Adding a dependency on a jar fetched that way to a BUILD file is all the user needs to do, right?

Re: jdeps. In this case, we're dealing with scala sources so calling `java_common.compile` won't work. I think there's a ticket for java_common API on bazel-side but can't find it at the moment. Can you think of any other workaround? Could the intellij plugin treat scala rules in some special way?

cc @ittaiz who is using the intellij+scala rules extensively

**Comment by brendandouglas** [comment]
We could certainly treat scala rules specially (and really need to, if they're not providing jdeps information). 

Removing the jdeps-based classpath minimization could seriously hurt performance for larger projects though, so a bazel-side fix would be preferable.

Filed https://github.com/bazelbuild/bazel/issues/4432

**Comment by ittaiz** [comment]
ok, got a chance to catch up on the thread.
The situation (on the rules scala side) is as follows- As Gregorz said when
compiling scala files we don't generate jdeps files. We have done work
similar to strict-java-deps but it still doesn't apply to unused-deps and
it doesn't use the same format as in jdeps (not sure it's ideal and not
sure it's documented as an API).
rules_scala actually integrates (for the most part) with the plugin
directly via a custom attribute it exposes on the output struct called
"scala" which AFAIK only the plugin reads.
I think this is unrelated to the subject here but I'll add that we
currently don't expose ijars to the plugin in that attribute since it's
unclear how to correlate the exposed ijars to the full jars (if this is
indeed unrelated then probably best not to sidetrack the issue to this
point).

**Comment by gkossakowski** [comment]
thanks attai for dumping the status! have you ran into the similar issue like described in the ticket? do you use scrooge_scala_library?

@brendandouglas are you aware of the special code-paths for the scala rules in the plugin? would that suggest some kind of a workaround while we think of a proper fix?
I'm wondering if jdeps optimization could be easily disabled with either a flag or simple code modification. This issue is currently blocking our migration to the plugin and I'm trying to figure out the next steps for the migration.

**Comment by brendandouglas** [comment]
OK, had another look at this. Lack of jdeps (and classpath minimization) for scala rules is a separate issue.

The actual problem here is that the scala-specific code isn't recognizing that this scala_library has no non-generated scala sources (and so it's output jars should be added to the project libraries).

We have this logic in our java-specific code -- I'll just add it to our scala code as well.

**Comment by gkossakowski** [comment]
Thank you! Excellent news that this can be fixed in the plugin.

**Comment by gkossakowski** [comment]
@brendandouglas  how much work would it be to fix this?
we're currently telling people to add whitespace `BUILD` files to side-step this issue but it's confusing especially for new users

**Comment by nevillelyh** [comment]
We're seeing the same issue, but with generated protobuf and avro files.
Using full target label i.e. `//.../avro:avro` instead of just `//.../avro` works around it but agree not idea.
@gkossakowski what's this whitespace `BUILD` file trick you mentioned?

**Comment by brendandouglas** [comment]
This was fixed upstream 2 weeks ago (along with the original diagnosis). Now it's just waiting on the next build, scheduled for release early next week.
